File: kotlin/everybody_codes/src/main/kotlin/org/gristle/pdxpuzzles/everybodycodes/y2025/Y25D13.kt
package org.gristle.pdxpuzzles.everybodycodes.y2025

import org.gristle.pdxpuzzles.everybodycodes.utilities.Day
import org.gristle.pdxpuzzles.utilities.math.isEven
import org.gristle.pdxpuzzles.utilities.parsing.getInts
import java.util.ArrayDeque

object Y25D13 : Day {
    private fun solve(ranges: Sequence<IntRange>, totalTurns: Long): Int {
        val lock = ArrayDeque<Pair<IntRange, Boolean>>(500)
        lock.addLast(1..1 to true)
        var start = 0
        var total = 1

        for ((idx, rng) in ranges.withIndex()) {
            total += rng.last - rng.first + 1
            if (idx.isEven()) {
                lock.addLast(rng to true)
            } else {
                start++
                lock.addFirst(rng to false)
            }
        }

        val totalTurns = (totalTurns + 1) % total
        var turns = 0

        for (i in start..start + lock.size) {
            val (rng, isForward) = lock.elementAt(i % lock.size)
            turns += rng.last - rng.first + 1
            if (turns >= totalTurns) {
                val diff = (turns - totalTurns).toInt()
                return if (isForward) {
                    rng.last - diff
                } else {
                    rng.first + diff
                }
            }
        }
        throw IllegalStateException("Unreachable!")
    }

    override fun part1(input: String): Int {
        val ranges = input.getInts().map { it..it }
        return solve(ranges, 2025)
    }
    override fun part2(input: String): Int {
        val ranges = input.getInts().chunked(2).map { (a, b) -> a..b }
        return solve(ranges, 20_252_025)
    }
    override fun part3(input: String): Int {
        val ranges = input.getInts().chunked(2).map { (a, b) -> a..b }
        return solve(ranges, 202_520_252_025)
    }
}

fun main() = Day.runDay(Y25D13::class)

//    Quest 1: 353 (1ms)
//    Quest 2: 7613 (4ms)
//    Quest 3: 217823 (3ms)
//    Total time: 10ms

File: rust/everybody-codes/src/bin/2025/y25d13.rs
use std::{collections::VecDeque, ops::RangeInclusive};

use everybody_codes::utilities::inputs::get_event_inputs;
use itertools::Itertools;
use utilities::{parsing::get_numbers::ContainsNumbers, structs::stopwatch::{ReportDuration, Stopwatch}};

type Input<'a> = &'a str;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let (input1, input2, input3) = get_event_inputs(25, 13);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input1), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input2), stopwatch.lap().report());
    println!("3. {} ({})", part3(&input3), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn solve(ranges: impl Iterator<Item = RangeInclusive<u64>>, total_turns: u64) -> u64 {
    let mut lock: VecDeque<(RangeInclusive<u64>, bool)> = VecDeque::with_capacity(500);
    lock.push_back((1..=1, true));
    let mut start = 0;
    let mut total: u64 = 1;

    for (idx, rng) in ranges.enumerate() {
        total += rng.end() - rng.start() + 1;
        if idx & 1 == 0 {
            lock.push_back((rng, true));
        } else {
            start += 1;
            lock.push_front((rng, false))
        }      
    }

    let total_turns = (total_turns + 1) % total;
    let mut turns: u64 = 0;
    
    for i in start.. {
        let (rng, is_forward) = lock.get(i % lock.len()).unwrap();
        turns += rng.end() - rng.start() + 1;
        if turns >= total_turns {
            let diff = turns - total_turns;
            return if *is_forward {
                rng.end() - diff
            } else {
                rng.start() + diff
            }
        }
    }
    unreachable!()
}

fn part1(input: Input) -> u64 {
    let ranges = input.get_numbers().map(|n| n..=n);
    solve(ranges, 2025)
}

fn part2(input: Input) -> u64 {
    let ranges = input.get_numbers().tuples().map(|(a, b)| a..=b);
    solve(ranges, 20_252_025)
}

fn part3(input: Input) -> u64 {
    let ranges = input.get_numbers().tuples().map(|(a, b)| a..=b);
    solve(ranges, 202_520_252_025)
}

#[test]
fn default() {
    let (input1, input2, input3) = get_event_inputs(25, 13);
    assert_eq!(353, part1(&input1));
    assert_eq!(7613, part2(&input2));
    assert_eq!(217823, part3(&input3));
}

// Input parsed (28μs)
// 1. 353 (8μs)
// 2. 7613 (6μs)
// 3. 217823 (20μs)
// Total: 65μs
